pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,264 @@
+	kubequotacontroller "github.com/kcp-dev/kcp/pkg/reconciler/kubequota"
+)
+
+// PluginName is the name of this admission plugin.
+const PluginName = "KCPKubeResourceQuota"
+
+// Register registers this admission plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(config io.Reader) (admission.Interface, error) {
+			configuration, err := resourcequota.LoadConfiguration(config)
+			if err != nil {
+				return nil, err
+			}
+
+			if configuration != nil {
+				if errs := validation.ValidateConfiguration(configuration); len(errs) > 0 {
+					return nil, errs.ToAggregate()
+				}
+			}
+
+			return NewKubeResourceQuota(configuration), nil
+		},
+	)
+}
+
+// NewKubeResourceQuota returns a new KubeResourceQuota admission plugin.
+func NewKubeResourceQuota(config *resourcequotaapi.Configuration) *KubeResourceQuota {
+	return &KubeResourceQuota{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+
+		userSuppliedConfiguration: config,
+
+		delegates: map[logicalcluster.Name]*resourcequota.QuotaAdmission{},
+	}
+}
+
+// KubeResourceQuota is an admission plugin that handles quota per logical cluster.
+type KubeResourceQuota struct {
+	*admission.Handler
+
+	// Injected/set via initializers
+	clusterWorkspaceLister tenancylisters.ClusterWorkspaceLister
+	kubeClusterClient      kubernetes.ClusterInterface
+	kubeInformers          informers.SharedInformerFactory
+	quotaConfiguration     quota.Configuration
+
+	// Manually set
+	userSuppliedConfiguration *resourcequotaapi.Configuration
+
+	lock      sync.RWMutex
+	delegates map[logicalcluster.Name]*resourcequota.QuotaAdmission
+}
+
+// ValidateInitialization validates all the expected fields are set.
+func (k *KubeResourceQuota) ValidateInitialization() error {
+	if k.clusterWorkspaceLister == nil {
+		return fmt.Errorf("missing clusterWorkspaceLister")
+	}
+	if k.kubeClusterClient == nil {
+		return fmt.Errorf("missing kubeClusterClient")
+	}
+	if k.kubeInformers == nil {
+		return fmt.Errorf("missing kubeInformers")
+	}
+	if k.quotaConfiguration == nil {
+		return fmt.Errorf("missing quotaConfiguration")
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &KubeResourceQuota{}
+var _ = initializers.WantsKcpInformers(&KubeResourceQuota{})
+var _ = initializers.WantsKubeInformers(&KubeResourceQuota{})
+var _ = initializers.WantsKubeClusterClient(&KubeResourceQuota{})
+var _ = initializer.WantsQuotaConfiguration(&KubeResourceQuota{})
+
+// Validate gets or creates a resourcequota.QuotaAdmission plugin for the logical cluster in the request and then
+// delegates validation to it.
+func (k *KubeResourceQuota) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	var delegate *resourcequota.QuotaAdmission
+
+	k.lock.RLock()
+	delegate = k.delegates[cluster.Name]
+	k.lock.RUnlock()
```

Comment:
would move this happy path logic into getOrCreateDelegate

Reply:
Fixed in squash commit